neqo-transport/src/connection.rs

```diff
@@ -271,6 +291,48 @@ impl IdleTimeout {
     }
 }
 
+/// StateManagement manages whether we need to send HANDSHAKE_DONE and CONNECTION_CLOSE.
+/// Valid state transitions are:
+/// * Idle -> HandshakeDone: at the server when the handshake completes
+/// * HandshakeDone -> Idle: when a HANDSHAKE_DONE frame is sent
+/// * Idle/HandshakeDone -> ConnectionClose: when closing
+/// * ConnectionClose -> CloseSent: after sending CONNECTION_CLOSE
+/// * CloseSent -> ConnectionClose: any time a new CONNECTION_CLOSE is needed
+#[derive(Debug, Clone, PartialEq)]
+enum StateSignaling {
+    Idle,
+    HandshakeDone,
+    ConnectionClose,
+    CloseSent,
+}
+
+impl StateSignaling {
+    pub fn handshake_done(&mut self) {
+        *self = Self::HandshakeDone
```

Comment:
Can you add a debug assertion that we are in the right state? Also handle not transitioning from ConnectionClose and CloseSent into HandshakeDone for non-debug build.